# Name: Godfrey maina
# RegNo: 19/03627
# Python Data reading assignment

## Reading data using procedural Steps

u can use the csv module to read csv using python

In [ ]:
import csv 

In [1]:
file = open('data.csv')

In [2]:
type(file)

_io.TextIOWrapper

## Reading data using pandas

### Reading CSV files

In [2]:
import pandas as pd

df = pd.read_csv('data.csv')
df.head()

Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0

### Reading JSON File

In [4]:
df = pd.read_json('data.json')

In [5]:
df.head()

Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406

### Analyzing DataFrames with pandas

The head() method returns the headers and a specified number of rows, starting from the top.

if the number of rows is not specified, the head() method will return the top 5 rows.

In [7]:
df = pd.read_csv('data.csv')
print(df.head(8))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3


The tail() method returns the headers and a specified number of rows, starting from the bottom.

In [9]:
df.tail(8)

Duration  Pulse  Maxpulse  Calories
161        45     90       130     260.4
162        45     95       130     270.0
163        45    100       140     280.9
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

The info() function gives you more information about the data set.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


## Data Cleaning


Data cleaning means fixing bad data in your data set.
Bad data could be:

1.Empty cells
2.Data in wrong format
3.Wrong data
4.Duplicates

  dropna() method removes empty cells

 By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

In [17]:
df.dropna(inplace = True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 164 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  164 non-null    int64  
 1   Pulse     164 non-null    int64  
 2   Maxpulse  164 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 6.4 KB


we can see that we have 164 rows now instead of 169

## Replace Empty Values


Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

In [18]:
df = pd.read_csv('data.csv')
df.fillna(130, inplace = True)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  169 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


## Replace Using Mean, Median, or Mode

In [20]:
df = pd.read_csv('data.csv')

x = df["Calories"].mean()

df["Calories"].fillna(x, inplace = True)

In [21]:
df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

## Removing Duplicates

To discover duplicates, we can use the duplicated() method.

In [22]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
164    False
165    False
166    False
167    False
168    False
Length: 169, dtype: bool

## Removing Duplicates

To remove duplicates, use the drop_duplicates() method.

In [23]:
df.drop_duplicates(inplace = True)

## Data Correlations

The corr() method calculates the relationship between each column in your data set.

In [24]:
df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.162098  0.003578  0.921907
Pulse    -0.162098  1.000000  0.787035  0.015408
Maxpulse  0.003578  0.787035  1.000000  0.194031
Calories  0.921907  0.015408  0.194031  1.000000

The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.